File: src/ladym/code/indexer.py
"""The codebase indexing driver (ARCHITECTURE.md §7).

Walks a repo, hashes each source file (incremental skip on unchanged), parses with
tree-sitter, extracts symbols + cross-refs, and stores them as L2 semantic memories +
L4-style code_refs. Files without a grammar degrade to sliding-window line chunking.
"""

from __future__ import annotations

import hashlib
import time
from dataclasses import dataclass, field
from pathlib import Path

from ..config import Config
from ..schema import CodeSymbol, Layer, Memory, MemoryType
from ..storage.embeddings import EmbeddingProvider
from ..storage.store import SQLiteStore
from .languages import detect_language, get_parser, get_spec
from .symbol_graph import RawSymbol, build_refs, extract_symbols

# cheap gitignore-style skip: directory basenames to never descend into
_SKIP_DIRS = {
    ".git", ".hg", ".svn", ".venv", "venv", "env", "__pycache__", ".mypy_cache",
    ".ruff_cache", ".pytest_cache", "node_modules", "dist", "build", ".tox",
    ".eggs", "target", "Pods", "DerivedData",
}


@dataclass
class IndexReport:
    files_seen: int = 0
    files_indexed: int = 0
    files_skipped_unchanged: int = 0
    files_skipped_unsupported: int = 0
    symbols_written: int = 0
    refs_written: int = 0
    elapsed_ms: float = 0.0
    errors: list[str] = field(default_factory=list)


def _hash_bytes(data: bytes) -> str:
    return hashlib.blake2b(data, digest_size=16).hexdigest()


def _should_ignore(path: Path, ignore_globs: list[str]) -> bool:
    if any(part in _SKIP_DIRS for part in path.parts):
        return True
    name = path.name
    return any(pat.rstrip("*").rstrip("/") in name for pat in ignore_globs)


def index_codebase(
    root: Path,
    store: SQLiteStore,
    embedder: EmbeddingProvider,
    *,
    cfg: Config,
    workspace: str | None = None,
    force: bool = False,
    language_filter: list[str] | None = None,
) -> IndexReport:
    """Walk ``root`` and index every supported source file."""
    start = time.time()
    ws = workspace or cfg.workspace
    root = Path(root).resolve()
    report = IndexReport()
    icfg = cfg.code_index

    for path in _walk(root, icfg.extra_ignore_globs):
        report.files_seen += 1
        lang = detect_language(path)
        if lang is None:
            report.files_skipped_unsupported += 1
            continue
        if language_filter is not None and lang not in language_filter:
            continue
        try:
            data = path.read_bytes()
        except OSError as e:
            report.errors.append(f"{path}: {e}")
            continue
        h = _hash_bytes(data)
        rel = str(path.relative_to(root))
        if not force:
            prev = store.get_indexed_hash(rel)
            if prev == h:
                report.files_skipped_unchanged += 1
                continue

        module_name = _module_name(rel, lang)
        spec = get_spec(lang)
        syms: list[RawSymbol] = []
        try:
            parser = get_parser(lang)
            if parser is not None and spec.definition_kinds:
                tree = parser.parse(data)
                syms = extract_symbols(
                    tree, data, spec,
                    module_name=module_name,
                    file_path=rel,
                    max_body_lines=icfg.max_body_lines_per_symbol,
                )
        except Exception as e:  # parse error ⇒ graceful fallback
            report.errors.append(f"{path}: parse failed ({e}); using chunk fallback")
            syms = []

        if not syms and spec.fallback_chunk_lines:
            # grammar-less / parse-failed: chunk by line windows
            syms = _chunk_fallback(data, module_name, rel, lang, spec.fallback_chunk_lines)

        # write file-level memory (always)
        file_summary = _file_summary(rel, lang, syms)
        _put_file_memory(store, embedder, ws, rel, lang, file_summary)

        # write per-symbol memories + projections
        for sym in syms:
            _put_symbol_memory(store, embedder, ws, rel, lang, sym, module_name)

        # write intra-file refs (resolved to qualified names where possible)
        refs = build_refs(syms, rel)
        store.put_code_refs(refs)

        store.set_indexed(rel, h)
        report.files_indexed += 1
        report.symbols_written += len(syms)
        report.refs_written += len(refs)

    report.elapsed_ms = (time.time() - start) * 1000.0
    return report


def _walk(root: Path, ignore_globs: list[str]):
    for p in sorted(root.rglob("*")):
        if not p.is_file():
            continue
        if _should_ignore(p, ignore_globs):
            continue
        yield p


def _module_name(rel: str, lang: str) -> str:
    base = rel.replace("/", ".").replace("\\", ".")
    for ext in (".py", ".js", ".ts", ".go", ".rs", ".java", ".c", ".cpp", ".rb", ".cs"):
        if base.endswith(ext):
            base = base[: -len(ext)]
            break
    if base.endswith(".index"):
        base = base[: -len(".index")]
    return base or "module"


def _file_summary(rel: str, lang: str, syms: list[RawSymbol]) -> str:
    if not syms:
        return f"{lang} source file: {rel} (no symbols extracted)"
    kinds: dict[str, int] = {}
    for s in syms:
        kinds[s.kind] = kinds.get(s.kind, 0) + 1
    kind_str = ", ".join(f"{n} {k}" for k, n in sorted(kinds.items()))
    top = ", ".join(s.name for s in syms[:8])
    return f"{rel} ({lang}): {kind_str}. Top symbols: {top}."


def _put_file_memory(
    store: SQLiteStore, embedder: EmbeddingProvider, ws: str,
    rel: str, lang: str, summary: str,
) -> None:
    # delete prior file memory for this path before re-inserting
    store.conn.execute(
        "DELETE FROM memories WHERE type = ? AND source = ? AND workspace = ?",
        (MemoryType.CODE_FILE.value, rel, ws),
    )
    mem = Memory(
        layer=Layer.SEMANTIC,
        type=MemoryType.CODE_FILE,
        content=summary,
        summary=summary[:120],
        tags=["code", lang],
        metadata={"file_path": rel, "language": lang},
        source=rel,
        workspace=ws,
        content_hash=_hash_bytes(summary.encode()),
    )
    store.put_memory(mem, vector=embedder.embed(summary))


def _put_symbol_memory(
    store: SQLiteStore, embedder: EmbeddingProvider, ws: str,
    rel: str, lang: str, sym: RawSymbol, module_name: str,
) -> None:
    # delete prior symbol memory with the same qualified_name in this workspace
    store.conn.execute(
        "DELETE FROM memories WHERE type = ? AND workspace = ? "
        "AND id IN (SELECT memory_id FROM code_symbols WHERE qualified_name = ?)",
        (MemoryType.CODE_SYMBOL.value, ws, sym.qualified_name),
    )
    content = _render_symbol_content(sym)
    mem = Memory(
        layer=Layer.SEMANTIC,
        type=MemoryType.CODE_SYMBOL,
        content=content,
        summary=f"{sym.kind} {sym.qualified_name}",
        tags=["code", lang, sym.kind],
        metadata={
            "file_path": rel,
            "language": lang,
            "qualified_name": sym.qualified_name,
            "kind": sym.kind,
        },
        source=rel,
        workspace=ws,
        content_hash=_hash_bytes(content.encode()),
    )
    vec = embedder.embed(content)
    store.put_memory(mem, vector=vec)
    store.put_code_symbol(
        CodeSymbol(
            memory_id=mem.id,
            file_path=rel,
            symbol_kind=sym.kind,
            qualified_name=sym.qualified_name,
            signature=sym.signature,
            docstring=sym.docstring,
            line_start=sym.line_start,
            line_end=sym.line_end,
            language=lang,
        )
    )


def _render_symbol_content(sym: RawSymbol) -> str:
    parts = [f"{sym.kind} {sym.qualified_name}"]
    if sym.signature:
        parts.append(f"signature: {sym.signature}")
    if sym.docstring:
        parts.append(f"doc: {sym.docstring}")
    if sym.body:
        parts.append("body:\n" + sym.body)
    return "\n".join(parts)


def _chunk_fallback(
    data: bytes, module_name: str, rel: str, lang: str, chunk_lines: int,
) -> list[RawSymbol]:
    """No-grammar fallback: split file into ``chunk_lines``-line windows."""
    try:
        text = data.decode("utf-8", errors="replace")
    except Exception:
        return []
    lines = text.split("\n")
    out: list[RawSymbol] = []
    for i in range(0, len(lines), chunk_lines):
        chunk = lines[i: i + chunk_lines]
        out.append(
            RawSymbol(
                kind="chunk",
                name=f"lines_{i + 1}",
                qualified_name=f"{module_name}.lines_{i + 1}",
                signature="",
                docstring="",
                line_start=i + 1,
                line_end=i + len(chunk),
                body="\n".join(chunk),
                calls=[],
            )
        )
    return out

File: code/indexer.go
package code

import (
	"encoding/hex"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/ProjAnvil/LadyM/config"
	"github.com/ProjAnvil/LadyM/schema"
	"github.com/ProjAnvil/LadyM/storage"
	"golang.org/x/crypto/blake2b"
)

// IndexReport reports the outcome of an indexing pass.
type IndexReport struct {
	FilesSeen               int
	FilesIndexed            int
	FilesSkippedUnchanged   int
	FilesSkippedUnsupported int
	SymbolsWritten          int
	RefsWritten             int
	ElapsedMs               float64
	Errors                  []string
}

var skipDirs = map[string]bool{
	".git": true, ".hg": true, ".svn": true, ".venv": true, "venv": true, "env": true,
	"__pycache__": true, ".mypy_cache": true, ".ruff_cache": true, ".pytest_cache": true,
	"node_modules": true, "dist": true, "build": true, ".tox": true, ".eggs": true,
	"target": true, "Pods": true, "DerivedData": true,
}

func hashBytes(data []byte) string {
	h, _ := blake2b.New(16, nil)
	h.Write(data)
	return hex.EncodeToString(h.Sum(nil))
}

func shouldIgnore(path string, ignoreGlobs []string) bool {
	parts := strings.Split(filepath.ToSlash(path), "/")
	for _, p := range parts {
		if skipDirs[p] {
			return true
		}
	}
	name := filepath.Base(path)
	for _, pat := range ignoreGlobs {
		stripped := strings.TrimRight(strings.TrimRight(pat, "*"), "/")
		if strings.Contains(name, stripped) {
			return true
		}
	}
	return false
}

// IndexCodebase walks root and indexes every supported source file.
//
// The whole pass runs under a cross-process flock on <db>.index.lock (mirrors
// the Python indexer): a second concurrent index — from another CLI, MCP
// server, or worker — fails fast with IndexInProgressError instead of
// interleaving writes.
func IndexCodebase(root string, store *storage.SQLiteStore, embedder storage.EmbeddingProvider, cfg *config.Config, workspace string, force bool, languageFilter []string) (*IndexReport, error) {
	release, err := acquireIndexLock(store.DBPath)
	if err != nil {
		return nil, err
	}
	defer release()

	start := time.Now()
	ws := workspace
	if ws == "" {
		ws = cfg.Workspace
	}
	absRoot, err := filepath.Abs(root)
	if err != nil {
		return nil, err
	}
	report := &IndexReport{}
	icfg := cfg.CodeIndex

	paths, err := walkFiles(absRoot, icfg.ExtraIgnoreGlobs)
	if err != nil {
		return nil, err
	}
	sort.Strings(paths)

	for _, path := range paths {
		report.FilesSeen++
		lang := DetectLanguage(path)
		if lang == "" {
			report.FilesSkippedUnsupported++
			continue
		}
		if languageFilter != nil && !contains(languageFilter, lang) {
			continue
		}
		data, err := os.ReadFile(path)
		if err != nil {
			report.Errors = append(report.Errors, path+": "+err.Error())
			continue
		}
		h := hashBytes(data)
		rel, err := filepath.Rel(absRoot, path)
		if err != nil {
			rel = path
		}
		rel = filepath.ToSlash(rel)
		if !force {
			prev, err := store.GetIndexedHash(rel)
			if err != nil {
				return nil, err
			}
			if prev == h {
				report.FilesSkippedUnchanged++
				continue
			}
		}

		moduleName := moduleName(rel, lang)
		spec := GetSpec(lang)
		syms, err := ExtractSymbols(data, lang, moduleName, rel, icfg.MaxBodyLinesPerSymbol)
		if err != nil { // parse error ⇒ graceful fallback (mirrors Python)
			report.Errors = append(report.Errors, fmt.Sprintf("%s: parse failed (%v); using chunk fallback", path, err))
		}

		if len(syms) == 0 && spec.FallbackChunkLines > 0 {
			syms = chunkFallback(data, moduleName, rel, lang, spec.FallbackChunkLines)
		}

		fileSummary := fileSummary(rel, lang, syms)
		if err := putFileMemory(store, embedder, ws, rel, lang, fileSummary); err != nil {
			return nil, err
		}
		for _, sym := range syms {
			if err := putSymbolMemory(store, embedder, ws, rel, lang, sym); err != nil {
				return nil, err
			}
		}
		refs := BuildRefs(syms, rel)
		codeRefs := make([]*schema.CodeRef, 0, len(refs))
		for _, r := range refs {
			codeRefs = append(codeRefs, &schema.CodeRef{SrcSymbol: r.SrcSymbol, DstSymbol: r.DstSymbol, RefKind: r.RefKind})
		}
		if err := store.PutCodeRefs(codeRefs); err != nil {
			return nil, err
		}

		if err := store.SetIndexed(rel, h, 0); err != nil {
			return nil, err
		}
		report.FilesIndexed++
		report.SymbolsWritten += len(syms)
		report.RefsWritten += len(refs)
	}

	report.ElapsedMs = float64(time.Since(start).Nanoseconds()) / 1e6
	return report, nil
}

func contains(list []string, s string) bool {
	for _, x := range list {
		if x == s {
			return true
		}
	}
	return false
}

func walkFiles(root string, ignoreGlobs []string) ([]string, error) {
	var out []string
	err := filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil // skip unreadable entries
		}
		if d.IsDir() {
			// Directories are pruned only via the built-in skipDirs set,
			// mirroring Python's `_walk` (extra globs match file basenames only).
			if shouldIgnore(path, nil) && path != root {
				return filepath.SkipDir
			}
			return nil
		}
		if shouldIgnore(path, ignoreGlobs) {
			return nil
		}
		out = append(out, path)
		return nil
	})
	return out, err
}

func moduleName(rel, lang string) string {
	base := strings.ReplaceAll(strings.ReplaceAll(rel, "/", "."), "\\", ".")
	for _, ext := range []string{".py", ".js", ".ts", ".go", ".rs", ".java", ".c", ".cpp", ".rb", ".cs"} {
		if strings.HasSuffix(base, ext) {
			base = base[:len(base)-len(ext)]
			break
		}
	}
	if strings.HasSuffix(base, ".index") {
		base = base[:len(base)-len(".index")]
	}
	if base == "" {
		return "module"
	}
	return base
}

func fileSummary(rel, lang string, syms []RawSymbol) string {
	if len(syms) == 0 {
		return lang + " source file: " + rel + " (no symbols extracted)"
	}
	kinds := map[string]int{}
	for _, s := range syms {
		kinds[s.Kind]++
	}
	kindKeys := make([]string, 0, len(kinds))
	for k := range kinds {
		kindKeys = append(kindKeys, k)
	}
	sort.Strings(kindKeys)
	var kindParts []string
	for _, k := range kindKeys {
		kindParts = append(kindParts, itoa(kinds[k])+" "+k)
	}
	top := make([]string, 0, 8)
	for i, s := range syms {
		if i >= 8 {
			break
		}
		top = append(top, s.Name)
	}
	return rel + " (" + lang + "): " + strings.Join(kindParts, ", ") + ". Top symbols: " + strings.Join(top, ", ") + "."
}

func putFileMemory(store *storage.SQLiteStore, embedder storage.EmbeddingProvider, ws, rel, lang, summary string) error {
	if _, err := store.DB().Exec(
		"DELETE FROM memories WHERE type = ? AND source = ? AND workspace = ?",
		string(schema.TypeCodeFile), rel, ws); err != nil {
		return err
	}
	m := schema.NewMemory(schema.LayerSemantic, schema.TypeCodeFile)
	m.Content = summary
	m.Summary = truncateStr(summary, 120)
	m.Tags = []string{"code", lang}
	m.Metadata = map[string]any{"file_path": rel, "language": lang}
	m.Source = rel
	m.Workspace = ws
	m.ContentHash = hashBytes([]byte(summary))
	vec, err := embedder.Embed(summary)
	if err != nil {
		return err
	}
	return store.PutMemory(m, vec)
}

func putSymbolMemory(store *storage.SQLiteStore, embedder storage.EmbeddingProvider, ws, rel, lang string, sym RawSymbol) error {
	if _, err := store.DB().Exec(
		"DELETE FROM memories WHERE type = ? AND workspace = ? AND id IN (SELECT memory_id FROM code_symbols WHERE qualified_name = ?)",
		string(schema.TypeCodeSymbol), ws, sym.QualifiedName); err != nil {
		return err
	}
	content := renderSymbolContent(sym)
	m := schema.NewMemory(schema.LayerSemantic, schema.TypeCodeSymbol)
	m.Content = content
	m.Summary = sym.Kind + " " + sym.QualifiedName
	m.Tags = []string{"code", lang, sym.Kind}
	m.Metadata = map[string]any{
		"file_path": rel, "language": lang,
		"qualified_name": sym.QualifiedName, "kind": sym.Kind,
	}
	m.Source = rel
	m.Workspace = ws
	m.ContentHash = hashBytes([]byte(content))
	vec, err := embedder.Embed(content)
	if err != nil {
		return err
	}
	if err := store.PutMemory(m, vec); err != nil {
		return err
	}
	return store.PutCodeSymbol(&schema.CodeSymbol{
		MemoryID: m.ID, FilePath: rel, SymbolKind: sym.Kind,
		QualifiedName: sym.QualifiedName, Signature: sym.Signature,
		Docstring: sym.Docstring, LineStart: sym.LineStart, LineEnd: sym.LineEnd,
		Language: lang,
	})
}

func renderSymbolContent(sym RawSymbol) string {
	parts := []string{sym.Kind + " " + sym.QualifiedName}
	if sym.Signature != "" {
		parts = append(parts, "signature: "+sym.Signature)
	}
	if sym.Docstring != "" {
		parts = append(parts, "doc: "+sym.Docstring)
	}
	if sym.Body != "" {
		parts = append(parts, "body:\n"+sym.Body)
	}
	return strings.Join(parts, "\n")
}

func chunkFallback(data []byte, moduleName, rel, lang string, chunkLines int) []RawSymbol {
	text := strings.ReplaceAll(string(data), "\r\n", "\n")
	lines := strings.Split(text, "\n")
	var out []RawSymbol
	for i := 0; i < len(lines); i += chunkLines {
		end := i + chunkLines
		if end > len(lines) {
			end = len(lines)
		}
		chunk := lines[i:end]
		out = append(out, RawSymbol{
			Kind: "chunk", Name: "lines_" + itoa(i+1),
			QualifiedName: moduleName + ".lines_" + itoa(i+1),
			LineStart:     i + 1, LineEnd: end,
			Body: strings.Join(chunk, "\n"),
			// Calls stays nil: Python's `_chunk_fallback` sets calls=[],
			// so chunked files produce no code_refs.
		})
	}
	return out
}

func truncateStr(s string, n int) string {
	r := []rune(s)
	if len(r) <= n {
		return s
	}
	return string(r[:n])
}

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	var buf [20]byte
	i := len(buf)
	for n > 0 {
		i--
		buf[i] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		i--
		buf[i] = '-'
	}
	return string(buf[i:])
}
